pkg/storage/ec/client.go

```diff
@@ -126,6 +126,13 @@ func (ec *ecClient) Put(ctx context.Context, nodes []*pb.Node, rs eestream.Redun
 		}
 	}
 
+	/* clean up the partially uploaded segment's pieces */
+	go func() {
+		<-ctx.Done()
+		ctx = context.Background()
+		_ = ec.Delete(ctx, successfulNodes, pieceID)
```

Comment:
Is this thread-safe?

Reply:
Also, the deletion on completion doesn't sound like the right thing here. Check whether it was actually canceled.

Reply:
Awesome... Egon... I had the same doubts, how to verify the CTRL+C actually cancelled all the way into the downstream. So what I did was when CTRL+C is received I didn't run the canel() function  and  added few debug info to see if ctx.Done() in line 131 above is executed. And it never ran, as ctx.Done is received only after cancel() is executed. This convinced me of two things that you just asked. 1. Is this thread-safe. and per my understanding and reading about context, it is thread safe. 2. is it really cancelled the previous request, and it did. 

Two things convinced me. 
1. without creating a new context as in line 132, I was not able to delete ALL the partial segments. I could delete partially the partial segments because from the time the user issued CTRL+C and the actual cancel propagated all the downstream, I could delete ONLY few segments as when cancel reached the downstream, it closed all the gRPC connections and it kicked me out. 

2. Taking to @jtolds , advised to create a new context and use it to clean up. When i did this, it did the trick, of being thread safe and as cancel is handled separately (created a new context and thats why thread safe), it cleaned up the whole partially uploaded segments and partially uploaded pieceIDs

Since I had the same questions and the logic/advice I used/followed answered my questions. I would like to hear your thoughts on this. Please let me know... hope I am not way off..

Reply:
This goroutine should be removed. Deleting pieces, which upload is currently in progress, should already be handled in the EncodeReader: https://github.com/storj/storj/blob/master/pkg/eestream/encode.go#L312

If this is not the case for some reason, it should be fixed there.

Reply:
I think this is deleting pieces that have been successfully uploaded.

Reply:
oh, you are right. My bad.

Reply:
OK. So this goroutine is started after the upload is finished (successfully or not). Why executing this logic in a goroutine? We can just check if the context has been cancelled and delete the pieces in the current goroutine.

Reply:
Changing above like to defer rather than go routine...

	/* clean up the partially uploaded segment's pieces */
	defer func() {
		select {
		case <-ctx.Done():
			ctx = context.Background()
			_ = ec.Delete(ctx, successfulNodes, pieceID)
		default:
		}
	}()

seems like this is a better choice as it waits for all the  background things to settle and clean up the last segment's pieceIDs before exiting... anythoughts? I tested with the above code and I dont see any difference and also will be bit more sequential rather than concurrent.

Reply:
I did some more extensive testing using goroutine versus defer and found that using grouting leaves more piecesIDs from cleaning up than with defer. Question is WHY? 

Reason: Seems like go routine kicks in as soon as CTRL+C is received and at that time the shared variable successfulNodes may have been updated after the go routine started. But if defer is used, runs only after everything is settled in more like a sequential than concurrent. 

Test Result: When goroutine was used, I saw quite a few pieceIDs got left over, but with defer, I saw only one pieceID got left over. This is with trying to upload a file 5.6GB size, and cancelled at 5.2GB-5.3GB range.